src/VisualStudio/Core/Def/Implementation/FindUsages/VisualStudioFindSymbolMonikerUsagesService.cs

```diff
@@ -0,0 +1,121 @@
+using Microsoft.CodeAnalysis.Editor.LanguageServiceIndexFormat;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Newtonsoft.Json.Linq;
+using VS.IntelliNav.Contracts;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.FindUsages
+{
+    [ExportWorkspaceService(typeof(IFindSymbolMonikerUsagesService), layer: ServiceLayer.Host), Shared]
+    internal partial class VisualStudioFindSymbolMonikerUsagesService : AbstractFindSymbolMonikerUsagesService
+    {
+        private readonly ICodeIndexProvider? _codeIndexProvider;
+
+        private readonly object _gate = new object();
+        private CancellationTokenSource? _lastNavigationCancellationSource;
+
+        [ImportingConstructor]
+        public VisualStudioFindSymbolMonikerUsagesService(
+            [Import(AllowDefault = true)] ICodeIndexProvider? codeIndexProvider)
+        {
+            _codeIndexProvider = codeIndexProvider;
+        }
+
+        public override async IAsyncEnumerable<ExternalReferenceItem> FindReferencesByMoniker(
+            DefinitionItem definition, ImmutableArray<SymbolMoniker> monikers,
+            IStreamingProgressTracker progress, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            if (_codeIndexProvider == null)
+                yield break;
+
+            var convertedMonikers = ConvertMonikers(monikers);
+            var currentPage = 0;
+            while (true)
+            {
+                var referenceItems = await FindReferencesByMonikerAsync(
+                    _codeIndexProvider, definition, convertedMonikers, progress, currentPage, cancellationToken).ConfigureAwait(false);
+
+                // If we got no items, we're done.
+                if (referenceItems.Length == 0)
+                    break;
+
+                foreach (var item in referenceItems)
+                    yield return item;
+
+                // Otherwise, we got some items.  Return them to our caller and attempt to retrieve
+                // another page.
+                currentPage++;
+            }
+        }
+
+        private async Task<ImmutableArray<ExternalReferenceItem>> FindReferencesByMonikerAsync(
+            ICodeIndexProvider codeIndexProvider, DefinitionItem definition, ImmutableArray<ISymbolMoniker> monikers,
+            IStreamingProgressTracker progress, int pageIndex, CancellationToken cancellationToken)
+        {
+            // Let the find-refs window know we have outstanding work
+            await using var _1 = await progress.AddSingleItemAsync().ConfigureAwait(false);
+
+            var results = await codeIndexProvider.FindReferencesByMonikerAsync(
+                monikers, includeDeclaration: false, pageIndex: pageIndex, cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
If this hits a network failure, whose responsibility is it to handle it?

Reply:
@bertanaygun?

Reply:
(sorry this one was lost amongst other notifications)
In our LSP implementation, we throw in case of network/auth issues since it provides some meaningful context to LSP command handler indicating call failed.

If Roslyn can't do anything with the exception though (since we don't have a retry UI logic), we can return an empty array instead of throwing. We would log the exceptions telemetry and also in output window (mainly for diagnostics).

I am fine either way, but it is probably better for Roslyn to catch exceptions and ignore them for now..

Reply:
roslyn catches and logs exceptions here now.